vm-virtio/src/vhost_user/net.rs

```diff
@@ -0,0 +1,313 @@
+// Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+use libc;
+use libc::EFD_NONBLOCK;
+use std::cmp;
+use std::io::Write;
+use std::os::unix::io::AsRawFd;
+use std::sync::atomic::AtomicUsize;
+use std::sync::Arc;
+use std::thread;
+use std::vec::Vec;
+
+use crate::VirtioInterrupt;
+use net_util::{MacAddr, MAC_ADDR_LEN};
+
+use vm_memory::{Address, Error as MmapError, GuestMemory, GuestMemoryMmap, GuestMemoryRegion};
+use vmm_sys_util::eventfd::EventFd;
+
+use super::super::{ActivateError, ActivateResult, Queue, VirtioDevice, VirtioDeviceType};
+use super::handler::VhostUserEpollHandler;
+use super::{Error, Result};
+use vhost_rs::vhost_user::{Master, VhostUserMaster};
+use vhost_rs::{VhostBackend, VhostUserMemoryRegionInfo, VringConfigData};
+use virtio_bindings::virtio_net;
+
+const VIRTIO_RING_F_EVENT_IDX: ::std::os::raw::c_uint = 29;
+const VIRTIO_F_NOTIFY_ON_EMPTY: ::std::os::raw::c_uint = 24;
+const VIRTIO_F_VERSION_1: ::std::os::raw::c_uint = 32;
+const VHOST_USER_F_PROTOCOL_FEATURES: ::std::os::raw::c_uint = 30;
+
+pub struct Net {
+    vhost_user_net: Master,
+    kill_evt: EventFd,
+    avail_features: u64,
+    acked_features: u64,
+    backend_features: u64,
+    config_space: Vec<u8>,
+    queue_sizes: Vec<u16>,
+}
+
+impl Net {
+    /// Create a new vhost-user-net device
+    pub fn new(
+        mac_addr: MacAddr,
+        path: &str,
+        num_queues_pairs: usize,
+        queue_size: u16,
+    ) -> Result<Net> {
+        let num_queues = 2 * num_queues_pairs;
+        let mut vhost_user_net =
+            Master::connect(path, num_queues as u64).map_err(Error::VhostUserCreateMaster)?;
+
+        let kill_evt = EventFd::new(EFD_NONBLOCK).map_err(Error::CreateKillEventFd)?;
+
+        let mut avail_features = 1 << virtio_net::VIRTIO_NET_F_GUEST_CSUM
+            | 1 << virtio_net::VIRTIO_NET_F_CSUM
+            | 1 << virtio_net::VIRTIO_NET_F_GUEST_TSO4
+            | 1 << virtio_net::VIRTIO_NET_F_GUEST_TSO6
+            | 1 << virtio_net::VIRTIO_NET_F_GUEST_ECN
+            | 1 << virtio_net::VIRTIO_NET_F_GUEST_UFO
+            | 1 << virtio_net::VIRTIO_NET_F_HOST_TSO4
+            | 1 << virtio_net::VIRTIO_NET_F_HOST_TSO6
+            | 1 << virtio_net::VIRTIO_NET_F_HOST_ECN
+            | 1 << virtio_net::VIRTIO_NET_F_HOST_UFO
+            | 1 << virtio_net::VIRTIO_NET_F_MRG_RXBUF
+            | 1 << VIRTIO_RING_F_EVENT_IDX
+            | 1 << VIRTIO_F_NOTIFY_ON_EMPTY
+            | 1 << VIRTIO_F_VERSION_1;
+
```

Comment:
Those avail_features should be gotten from vhost-user-net backend, and we do not need set here.  vhost user device only check the features bits according to cmdline settings and decide which feature will be supported.

Reply:
The vring features together with virtio device features, both need to be negotiation between VMM and vhost-user backend, since it's possible that some features backend support but VMM not, negotiation will be the safe way to get it.